primitives/src/runtime_api.rs

```diff
@@ -18,31 +18,97 @@
 //! of the Runtime API exposed from the Runtime to the Host.
 //!
 //! The functions in trait ParachainHost` can be part of the stable API
-//! (which is versioned) or they can be staging (aka unstable functions).
+//! (which is versioned) or they can be staging (aka unstable/testing
+//! functions).
 //!
-//! All stable API functions should use primitives from the latest version.
-//! In the time of writing of this document - this is v2. So for example:
-//! ```ignore
-//! fn validators() -> Vec<v2::ValidatorId>;
-//! ```
-//! indicates a function from the stable v2 API.
+//! The separation outlined above is achieved with the versioned api feature
+//! of `decl_runtime_apis!` and `impl_runtime_apis!`. Before moving on let's
+//! see a quick example about how api versioning works.
 //!
-//! On the other hand a staging function's name should be prefixed with
-//! `staging_` like this:
-//! ```ignore
-//! fn staging_get_disputes() -> Vec<(vstaging::SessionIndex, vstaging::CandidateHash, vstaging::DisputeState<vstaging::BlockNumber>)>;
+//! # Runtime api versioning crash course
+//!
+//! The versioning is achieved with the `api_version` attribute. It can be
+//! placed on:
+//! * trait declaration - represents the base version of the api.
+//! * method declaration (inside a trait declaration) - represents a versioned
+//!   method, which is not available in the base version.
+//! * trait implementation - represents which version of the api is being
+//!   implemented.
+//!
+//! Let's see a quick example:
+//!
+//! ```rust(ignore)
+//! sp_api::decl_runtime_apis! {
+//! 	#[api_version(2)]
+//! 	pub trait MyApi {
+//! 		fn fn1();
+//! 		fn fn2();
+//! 		#[api_version(3)]
+//! 		fn fn3();
+//! 		#[api_version(4)]
+//! 		fn fn4();
+//! 	}
+//! }
+//!
+//! struct Runtime {}
+//!
+//! sp_api::impl_runtime_apis! {
+//!     #[api_version(3)]
+//!     impl self::MyApi<Block> for Runtime {
+//!         fn fn1() {}
+//!         fn fn2() {}
+//!         fn fn3() {}
+//!     }
+//! }
 //! ```
+//! A new api named `MyApi` is declared with `decl_runtime_apis!`. The trait declaration
+//! has got an `api_version` attribute which represents its base version - 2 in this case.
```

Comment:
When will we modify the base version? Is it obsoleted by individual method versioning?

Reply:
I'm not sure I get the question correctly so please point me in the right direction :)

The base version is what we had in substrate before the versioned methods. It is not changed as a concept.
One can think about it as the first version which can be implemented or default version. It depends on the point of view.

When it is modified - in our case, when a new stable version is released. Once the base version is bumped for example from 2 to 3 one can no more implement v2 of the runtime api with `impl_runtime_apis!`.

Is this what you asked?

Reply:
Yup, that makes sense, thanks! I somehow thought that with impl versions and individual method versioning, there would be no need for trait declaration version. But enforcing minimal impl version makes sense.